Source/Control/adsr.cpp

```diff
@@ -3,79 +3,75 @@
 
 using namespace daisysp;
 
-float Adsr::Tau2Pole(float tau)
+
+void Adsr::Init(float sample_rate, int blockSize)
 {
-    return expf(-1.0f / (tau * sample_rate_));
+    sample_rate_ = sample_rate / blockSize;
+    SetTime(ADSR_SEG_ATTACK, 0.1f);
+    SetTime(ADSR_SEG_DECAY, 0.1f);
+    SetTime(ADSR_SEG_RELEASE, 0.1f);
+    sus_level_ = 0.7f;
+    x_         = 0.0f;
 }
 
-float Adsr::AdsrFilter()
+void Adsr::Retrigger(bool hard)
 {
-    y_ = b_ * x_ + a_ * y_;
-    return y_;
+    mode_ = ADSR_SEG_ATTACK;
+    if(hard)
+        x_ = 0.f;
 }
 
-
-void Adsr::Init(float sample_rate)
+void Adsr::SetTime(int seg, float time)
 {
-    seg_time_[ADSR_SEG_ATTACK]  = 0.1f;
-    seg_time_[ADSR_SEG_DECAY]   = 0.1f;
-    sus_                        = 0.7f;
-    seg_time_[ADSR_SEG_RELEASE] = 0.1f;
-    //timer_ = 0;
-    a_           = 0.0f;
-    b_           = 0.0f;
-    x_           = 0.0f;
-    y_           = 0.0f;
-    prev_        = 0.0f;
-    atk_time_    = seg_time_[ADSR_SEG_ATTACK] * sample_rate;
-    sample_rate_ = sample_rate;
+    if(seg_time_[seg] != time)
+    {
+        seg_time_[seg] = time;
+        if(time > 0.f)
+        {
+            const float attackTarget = logf(0.5f);
+            const float decayTarget  = logf(1. / M_E);
+            float       target
+                = (seg == ADSR_SEG_ATTACK) ? attackTarget : decayTarget;
+            seg_D0_[seg] = 1.f - expf(target / (time * sample_rate_));
+        }
+        else
+            seg_D0_[seg] = 1.f; // instant change
+    }
 }
 
+
 float Adsr::Process(bool gate)
 {
-    float pole, out;
-    out = 0.0f;
+    float out = 0.0f;
 
     if(gate && mode_ != ADSR_SEG_DECAY)
-    {
         mode_ = ADSR_SEG_ATTACK;
-        //timer_ = 0;
-        pole      = Tau2Pole(seg_time_[ADSR_SEG_ATTACK] * 0.6f);
-        atk_time_ = seg_time_[ADSR_SEG_ATTACK] * sample_rate_;
-        a_        = pole;
-        b_        = 1.0f - pole;
-    }
     else if(!gate && mode_ != ADSR_SEG_IDLE)
-    {
         mode_ = ADSR_SEG_RELEASE;
-        pole  = Tau2Pole(seg_time_[ADSR_SEG_RELEASE]);
-        a_    = pole;
-        b_    = 1.0f - pole;
-    }
-
-    x_    = (int)gate;
-    prev_ = (int)gate;
 
+    float D0(seg_D0_[mode_]);
```

Comment:
While playing with this PR a bit more, I noticed it does not reset this array when ADSR is idling. To have a proper retrigger, I think we should reset it as well.